include/mqtt/broker/security.hpp

```diff
@@ -0,0 +1,309 @@
+                authentication_.insert( { name, auth });
+            } else if(method == "client_cert") {
+                authentication auth(authentication::method::client_cert);
+                authentication_.insert({ name, auth });
+            } else if(method == "anonymous") {
+                if(anonymous) throw std::runtime_error("Only a single anonymous user can be configured, anonymous user: " + *anonymous);
+                anonymous = name;
+
+                authentication auth(authentication::method::anonymous);
+                authentication_.insert( { name, auth });
+            } else throw std::runtime_error("An invalid method was specified: " + method);
+        }
+        for(auto const& i: root.get_child("group")) {
+            std::string name = i.second.get<std::string>("name");
+            if(!is_valid_group_name(name)) throw std::runtime_error("An invalid group name was specified: " + name);
+
+            group group;
+            if(i.second.get_child_optional("members")) {
+                for(auto const& j: i.second.get_child("members")) {
+                    auto username = j.second.get_value<std::string>();
+                    if(!is_valid_user_name(username)) throw std::runtime_error("An invalid user name was specified: " + username);
+                    group.members.push_back(username);
+                }
+            }
+
+            groups_.insert({ name, group });
+        }
+
+        for(auto const& i: root.get_child("authorization")) {
+            std::string name = i.second.get<std::string>("topic");
+            //if(!is_valid_topic(name)) throw std::runtime_error("An invalid topic was specified: " + name);
+
+            auto type = get_auth_type(i.second.get<std::string>("type"));
+
+            authorization auth(name, type);
+            if(i.second.get_child_optional("sub")) {
+                for(auto const& j: i.second.get_child("sub"))
+                    auth.sub.push_back(j.second.get_value<std::string>());
+            }
+
+            if(i.second.get_child_optional("pub")) {
+                for(auto const& j: i.second.get_child("pub"))
+                    auth.pub.push_back(j.second.get_value<std::string>());
+            }
+
+            authorization_.insert({ name, auth });
+        }
+
+        hash_type = root.get<std::string>("config.hash");
+        salt = root.get<std::string>("config.salt");
+        validate();
+    }
+
+    authorization::type auth_pub(string_view const& topic, string_view const& username) const {
+        authorization::type result_type = authorization::type::deny;
+
+        auth_pub_map.find(topic, [&](std::string const &allowed_username, authorization::type type) {
+            if (allowed_username == username) result_type = type;
+        });
+
+        return result_type;
+    }
+
+    std::map<std::string, authorization::type> auth_sub(string_view const& topic) const {
```

Comment:
Good point. 
Why do you choose two phase checking?
I mean
Phase 1. getting allowed users list by topic
Phase 2. cheking deliver target user using the result of Phase1

Reply:
You do not know beforehand which topics are going to be published, so which topics would you use in the multi index ? Only when you get a specific topic at a publish, you can check if it is allowed to publish/subscribe this. 

I use two phases, because the subscription_map gives a list of sessions which are subscribed using a callback. The authorization map also gives a list of usernames which are allowed to subscribe. Only the intersection of these two lists, are sessions/usernames which need to get the message delived.

Reply:
allowed users: { user1, user2, user3 }
subscribed sessions: { user2, user4 }
publish: { user2 } (intersection)

Reply:
I think I don't understand the logic yet.

I mean

```cpp
        auto deliver =
            [&] (session_state& ss, subscription& sub) {

                // topic and username combination version
                auto allowed = security.auth_sub(topic, ss.get_username());
                if (allowed != security::authorization::type::allow) return;
```

deliver() is called ffrom `subs_map_.modify()`.

Reply:
I guess i do not understand what you are proposing now. 
You want to cache the results of authorization lookups in a multi-index ? Between deliver calls ?

Reply:
I'm just asking now. Authorization is for topic and username combination.
So the simple interface is `can_sub(topic, username)`.
I don't consiter performance yet.
This is simpler than two phase one.

Reply:
> unbounded different topics.
What does it mean? wildcard ?
I guess I don't understand unbound topics.

Reply:
No, simple interface is possible. But there is performance difference, if multiple users are allowed to publish same topic, you are looking up the same topic multiple times. 

With no wildcard, you can use multiindex
With wildcard, you have to use the subscription_map, to check for wildcard matching

Reply:
> No, simple interface is possible. But there is performance difference, if multiple users are allowed to publish same topic, you are looking up the same topic multiple times.

Maybe I don't understand this point. One on_publish sequence is for only one topic.
Two phase logic has some cache? (I couldn't find it so far)

I think that the individulal publish is independent.

Reply:
I got confused about the word publish. 
There are two different meanings.
One is client's publish. The other is broker's publish. I use the word publish as former meanning. And the latter, I use deliver.

Reply:
Yes, the list of usernames (auth_users) is the cache. 
https://github.com/redboltz/mqtt_cpp/pull/896/files#diff-8438e6574a9cbca06fc27190e2dab4239f23e5e53b736428ad7b1d37dec831e7R2153

Reply:
Is `auth_sub()` high cost function?

Reply:
How `auth_sub_map` , internal structure, manage wildcards?

Reply:
The class `multiple_subscription_map` is not only to manage subscription but also authorization?

Reply:
The internal data structure is the subscription_map. So instead of being subscribed based on a topic filter, the authorization used topic filters to store a list of users. The cost is not high for a lookup. 

But the auth_users requires heap allocation to store / cache the result. It builds a set of authorized usernames.

Reply:
Maybe I need to understand how `multiple_subscription_map` is used for authorization.
`multiple_subscription_map` is a trie.

```
root A1/B1/C1
           C2
        B2/C1
     A2
```

There are 8 nodes. Is that right?
8 nodes means as follows:
root, A1, A1/B1, A1/B1/C1, A1/B1/C2, A1/B2, A1/B2/C1, A2

And each node can have values.

For managing subscriptions, the values are session.
For managing authorizations, the values are username.

It that right?

Reply:
Yes, that is correct. 

Only thing is:
For managing authorizations, the values are authorized usernames and groupnames at the node.

Reply:
Thnaks !
If I understand correctly, `multiple_subscription_map` is topic oriented data structure. So there is no `find()` interface that has a pair of topic and username parameter. `multiple_subscription_map` treats widlcard well. Wildcard related search is required by both deliver and authorization. But they are independent *1.
Now, focusing on authorization, use topic as the key and got usernames as values. 
This is `auth_sub()`. And then call `auth_sub_user()` using gotten usernames.

This design is simple. The part of authorization is independent.
`auth_sub()` returns all matched user - persmissions for the parameter topic. Even if the user doesn't subscribe the topic. 
If there are permissiong `topic1` - `user1` , `topic1` - `user2` ... `topic1` - `user10000`. And only `user1` subscribes `topic1`. `auth_sub()` returns 10000 entries. `auth_sub_user()` is O(log10000) complexity (std::map case).
If there is a good way to limit seraching only subscribed user, then the logic could be more efficient but I'm not sure it is possible.

Reply:
> If there is a good way to limit seraching only subscribed user, then the logic could be more efficient but I'm not sure it is possible.

yes, one way is to start with the actual subscribed sessions. First get a list of sessions which are actually subscribed, and maybe remove them if they are not authorized. This could limit the number of usernames which are considered ?

Reply:
Hmm, I'm not sure. 
The current design (this PR) is simple because well modulized. Authorization and subscription independency is good. 

Is `auth_sub_map` inserted from json on boot ?
Perhaps `auth_sub_map_all` and `auth_sub_map_deliver` could be a solution.
On boot, create `auth_sub_map_all`. It is the same as the current `auth_sub_map`.
On subscribe, check permission using `auth_sub_map_all` and then, insert into `auth_sub_map_deliver`.
(I'm not sure how wildcard treat well)

And on deliver, use `auth_sub_map_deliver` instead of `auth_sub_map_all`.
What do you think?
I think that it is the best approach if it is possible.

I come up with another memo approach based on the current approach.
See https://wandbox.org/permlink/QudhQN523ntjdgBG
I don't think it is the best approach but it reduce the complexity to O(logN)  N is number of delivered users. It is at most number of subscribers.

Reply:
> Is `auth_sub_map` inserted from json on boot ?

yes, it is generated on boot only
maybe in future also on reloading the auth file if it was changed ?

Reply:
> On subscribe, check permission using `auth_sub_map_all` and then, insert into `auth_sub_map_deliver`.
> (I'm not sure how wildcard treat well)

No handling wildcards in this scenario is very difficult, because you can have wildcard subscription as well as in `auth_sub_map_all` entries. And also auth rules may override previous auth rule, if it is more specific. 

Same issue applies to detecting if user is allowed to subscribe. You should deny subscription if there are no valid topic at all which can be generated for the subscription.

Reply:
> maybe in future also on reloading the auth file if it was changed ?

I think so.

> No handling wildcards in this scenario is very difficult,

I don't understand in detail but I can imagine it is difficult.
So let's give up limit subscription only authorization.

Move on https://wandbox.org/permlink/QudhQN523ntjdgBG 
What do you think?

Reply:
> Move on https://wandbox.org/permlink/QudhQN523ntjdgBG
> What do you think?

This is what is implemented now, only thing is, result is not cached between delivers. But do you really want to do that, because this requires synchronisation between threads. The cache needs to be locked. And gets very big, if old topics are not removed.

Reply:
It is based on the PR (your PR). memo is new thing.
I added lock to memo. See https://wandbox.org/permlink/cbl3e4jLg3sSBSno
At the first deliver, 1 shared_lock and 1 lock_guard are required.
But since the second deliver, only 1 shared_lock is required. It can share multiple threads.

Reply:
RAM consumption is very small for me. But if someone care, so `--enable-authorization-cache` option.

Reply:
I really doubt this will give any performance benefit. The lookup in subscription map is very efficient. The only possible concern is the temporary set of usernames. 

Instead of std::string, we might store a list of userid in this. This removes the overhead of allocating string. Additionally, a boost flat_set can be used, instead of a set. Different solutions need to be benchmarked before making a decision. 

But first, the authorization should be correct. That is the most important. Otherwise this is a security issue.

Reply:
memo is not an essential part of authorization. So let's leave that.
Now, mqtt_cpp have bench tool. 

We can measure

1. No authorization (current master)
2. Authorization (this PR)
3. Authorization with memo (on multi-threaded) 

After implementation is finished , I will measure 1 and 2. If performance is seriously down, then try 3.

Reply:
By the way, over 400,000 publish per second observed (QoS0, 8 io_contexts, 4 threads for each io_contexts, 1kbytes payload).

Reply:
yes, I read you performed a benchmark. That sounds quite high, but I am not sure how this compares to mosquitto or activemq, emq for example?

Indeed, we need to compare performance with/without authorization. But performance is now also related to number of users in the system, that is something to keep in mind....

Reply:
> yes, I read you performed a benchmark. That sounds quite high, but I am not sure how this compares to mosquitto or activemq, emq for example?

I couldn't find any good measure tools.
I need pure publish performance measure tool, so I create it.
Publish should be started after all connect and subscribe are completed.
Some of brokers has authorization cache so idle publish (out of measure) should be supported.
Publish interval should be the same across different clients.

I mean:
1,000 cleints.
One client 10 publish / second

Total 10,000 publish / second
The publish interval across clients should be 1 / (1,000 * 10) sec = 100us
It is offtopic. Let's back to the topic.

> I don't think you will see much difference in performance with small number of users.

I agree. log2N is efficient. 1000 users for 1 topic requires about only 10  comparison.
But for each from client publish to the topic,  insert into the map to element 1000 times. If I understand correctly.
Anyway, let's measure after later.